internal/config/config.go

```diff
@@ -148,6 +149,29 @@ func parseParameters(parametersStr string) (Parameters, error) {
 		return Parameters{}, err
 	}
 
+	tokensJSON := params["csi.storage.k8s.io/serviceAccount.tokens"]
+	if tokensJSON != "" {
+		// The csi.storage.k8s.io/serviceAccount.tokens field is a JSON object
+		// marshalled into a string. The object keys are audience name (string)
+		// and the values are embedded objects with "token" and
+		// "expirationTimestamp" fields for the corresponding audience.
+		var tokens map[string]struct {
+			Token               string `json:"token"`
+			ExpirationTimestamp string `json:"expirationTimestamp"`
+		}
+		if err := json.Unmarshal([]byte(tokensJSON), &tokens); err != nil {
+			return Parameters{}, fmt.Errorf("failed to unmarshal service account tokens: %w", err)
+		}
+
+		audience := "vault"
+		if parameters.Audience != "" {
```

Comment:
For my own reference: the audience just needs to be a service account that a token will be generated for?

Reply:
The audience can be any arbitrary string, not necessarily a service account name. If the user wants to take advantage of CSI TokenRequest tokens from the driver, they'll need to set an audience that matches under `tokenRequests` in the CSI driver definition: https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/charts/secrets-store-csi-driver/templates/csidriver.yaml#L13 (i.e. `--set tokenRequests[0].audience="vault"` via helm). The driver will create and send a token for each entry in that list.

If there's no token matching our expected audience, we'll currently fall back to the existing behaviour of generating the token in the provider, but I'd like to deprecate that fall back when this driver feature is more stable.

The idea is then the consumer of the JWT (Vault in our case) will validate that the JWT was intended to be consumed by Vault - for Kubernetes auth, that expected audience is configurable per role: https://www.vaultproject.io/api-docs/auth/kubernetes#create-role.

I'll write a changelog note announcing the plan to deprecate as well as the feature itself.

Reply:
Cool, thanks for the information.

Reply:
One knock-on I only just thought of.. Defaulting to an audience that is not the Kubernetes' cluster's default audience means these tokens won't be usable for our "use the client's JWT for auth" k8s auth option: https://www.vaultproject.io/docs/auth/kubernetes#use-the-vault-client-s-jwt-as-the-reviewer-jwt

I think that's probably a good thing that's more in the spirit of JWTs/OIDC, because it means tokens are generated to be consumed by a single audience. And even if we did go down the default audience route, it wouldn't simplify the driver configuration - users would just have to do a slightly confusing `--set tokenRequests[0].audience=""`.

So if you agree, probably the only action item here is to document that subtle detail when we next release the CSI provider.